Add conformance tests to CAPV

aartij17 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide/pull-requests.md#the-pull-request-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
5. If this PR changes image versions, please title this PR "Bump <image name> from x.x.x to y.y.y."
-->

**What this PR does / why we need it**:
This PR adds conformance tests to CAPV. 



**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

srm09 commented on test/e2e/common.go:
```diff
@@ -17,11 +17,55 @@ limitations under the License.
 	By(fmt.Sprintf(format, a...))
 }
+
+func setupSpecNamespace(ctx context.Context, specName string, clusterProxy framework.ClusterProxy, artifactFolder string) (*corev1.Namespace, context.CancelFunc) {
```
Not necessary to bring this change in, use the CAPI E2E code as a library. Similar to the comment I have added below.

srm09 commented on test/e2e/config/vsphere-ci.yaml:
```diff
@@ -169,6 +169,8 @@ variables:
   INIT_WITH_KUBERNETES_VERSION: "v1.19.1"
   VSPHERE_INSECURE_CSI: "true"
+  KUBETEST_CONFIGURATION: "./data/kubetest/conformance.yaml"
+
```
nit: remove extra line

srm09 reviewed (changes requested):
Added inline comments to use CAPI e2e code as a library

srm09 commented on test/e2e/common.go:
```diff
@@ -17,9 +17,38 @@ limitations under the License.
 	"fmt"
 
 	. "github.com/onsi/ginkgo" // nolint:gci
+	corev1 "k8s.io/api/core/v1"
```
In general a comment about this file, any reason why we are moving the test vars to the `common.go` file. IMO the purpose of the file is to host constants and functions to be shared between different tests.

aartij17 commented:
@srm09 I had moved the constants to `common.go` cause I was reusing them in another file(`k8s_conformance.go`). Exported variables/functions declared in a `*_test.go` couldn't be used in another file. 
But now I removed `k8s_conformance.go`. I will move the variables back to the original file.

srm09 reviewed (commented):
Looks good to me except for a nit. Thanks for the great work.

srm09 commented:
/ok-to-test

srm09 commented on test/e2e/k8s_conformance_test.go:
```diff
@@ -0,0 +1,34 @@
+			BootstrapClusterProxy: bootstrapClusterProxy,
+			ArtifactFolder:        artifactFolder,
+			SkipCleanup:           skipCleanup,
+		}
```
Whoops, include the new flavor that was created for this particular test which provisions beefy k8s worker nodes to run the tests without a hitch.

srm09 commented on test/e2e/k8s_conformance_test.go:
```diff
@@ -0,0 +1,34 @@
+			ClusterctlConfigPath:  clusterctlConfigPath,
+			BootstrapClusterProxy: bootstrapClusterProxy,
+			ArtifactFolder:        artifactFolder,
+			SkipCleanup:           skipCleanup,
```
```suggestion
			SkipCleanup:           skipCleanup,
			Flavor:                       "conformance",
```

geetikabatra commented on test/e2e/data/infrastructure-vsphere/kustomization/conformance/worker-node-size.yaml:
```diff
@@ -0,0 +1,25 @@
+spec:
+  template:
+    spec:
+      cloneMode: linkedClone
```
Is this specific to Vsphere? @srm09 and @aartij17 ?

srm09 replied:
The template is specific to vSphere. Does that answer your question?

aartij17 commented:
/test pull-cluster-api-provider-vsphere-test

srm09 commented:
/lgtm
/approve
